core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -333,6 +347,122 @@ public long nextFetchOffset() {
+                        if (batchStartOffset >= logStartOffset) {
+                            break;
+                        }
+                        InFlightBatch inFlightBatch = entry.getValue();
+                        boolean fullMatch = inFlightBatch.firstOffset >= startOffset && inFlightBatch.lastOffset < logStartOffset;
+
+                        // Maintain state per offset if the inflight batch is not a full match or the offset state is managed.
+                        if (!fullMatch || inFlightBatch.offsetState != null) {
+                            log.debug("Subset or offset tracked batch record found while trying to update offsets and cached" +
+                                            " state map due to LSO movement, batch: {}, offsets to update - " +
+                                            "first: {}, last: {} for the share partition: {}-{}", inFlightBatch, startOffset,
+                                    logStartOffset - 1, groupId, topicIdPartition);
+
+                            if (inFlightBatch.offsetState == null) {
+                                if (inFlightBatch.batchState() != RecordState.AVAILABLE) {
+                                    continue;
+                                }
+                                inFlightBatch.maybeInitializeOffsetStateUpdate();
+                            }
+
+                            archivePerOffsetBatchRecords(inFlightBatch, startOffset, logStartOffset - 1, stateBatches);
+                        } else {
+                            // The in-flight batch is a full match hence change the state of the complete batch.
+                            archiveCompleteBatch(inFlightBatch, stateBatches);
+                        }
+                    }
+                    // If we have transitioned the state of any batch/offset from AVAILABLE to ARCHIVED,
+                    // then there is a chance that the next fetch offset can change.
+                    if (!stateBatches.isEmpty()) {
+                        findNextFetchOffset.set(true);
+                        // We will write the archived state batches to the persister lazily during acknowledge/release acquired records API call.
+                        pendingWriteStateBatches.addAll(stateBatches);
+                    }
+                    // The new startOffset will be the log start offset.
+                    startOffset = logStartOffset;
+                    if (endOffset < startOffset) {
+                        // This case means that the cached state is completely fresh now.
+                        endOffset = startOffset;
+                    }
+                }
+                // We will only write the start and end offset to the persister actively.
+                // Even if write share group state RPC call fails, we will not be rolling back startOffset and endOffset.
+                isWriteShareGroupStateSuccessful(new ArrayList<>());
```

Comment:
Are we updating the persister right away?

Reply:
we are only updating the start offset. The persister state batches will be written lazily during acknowledge or release acquired records.

Reply:
we will be removing it and let the next acknowledge call take care of writing the startOffset to the persister.